### PR TITLE
feat: expose `wasm-opt` optimizer in `ic-wasm` to users

### PR DESCRIPTION
**Pull request by kentosugama**
# Description
[LANG-124](https://dfinity.atlassian.net/browse/LANG-124)

[Profiling data](https://docs.google.com/document/d/1ICXF083-hfRZr2OfRUk8OxmZqYDQL4XINHQ0FXuDqTA/edit) suggests that it would be useful to take advantage of the performance and binary size wasm optimizations offered by [`wasm-opt`](https://github.com/WebAssembly/binaryen). 

See also results in the [canister profiling repo](https://github.com/dfinity/canister-profiling/pull/35).

This tool has been integrated with `ic-wasm` in this PR: https://github.com/dfinity/ic-wasm/pull/28. The tool is added under the `shrink` command that is currently used in `dfx` to perform binary size reduction of canisters.

The next step is to expose this feature to users by allowing users to opt into the optimizer and specify the optimization level in `dfx.json`. Then to invoke `ic-wasm` in `dfx` according to this argument, similarly to `shrink`.

Note: This tool has been used in the past to perform binary size reductions for canisters, but was replaced because it deletes the custom metadata sections from wasm modules. The `ic-wasm` feature invokes `wasm-opt` while preserving these sections.

Example usage in `dfx.json`:
```
"canisters" : {
  "backend" : {
    "optimize" : "cycles"
  }
}
```

# How Has This Been Tested?
The following are `build.bash` tests, ensuring that builds succeed with optimizations turned on, but that the feature is turned off by default, especially for custom canisters.

```
@test "build succeeds if enable optimize" {
    jq '.canisters.e2e_project_backend.optimize="cycles"' dfx.json | sponge dfx.json
    dfx_start
    dfx canister create --all
    assert_command dfx build
}

@test "build custom canister default no optimize" {
    install_asset custom_canister
    install_asset wasm/identity

    dfx_start
    dfx canister create --all
    assert_command dfx build custom
    assert_not_match "Optimize"

    jq '.canisters.custom.optimize="size"' dfx.json | sponge dfx.json
    assert_command dfx build custom
    assert_match "Optimize"
}
```

The following is a test that ensures that `dfx start` correctly rejects invalid optimization options in `dfx.json`.
```
@test "invalid optimization level" {
    jq '.canisters.bad_optimization_level.optimize="bad_level"' dfx.json | sponge dfx.json
    assert_command_fail dfx_start
    assert_match "expected one of "
}
```

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.
- [x] Add e2e tests
- [x] Add e2e error tests
- [x] Update ic-wasm dependency
- [x] Add default "cycles" and "size" options
- [x] Error handle invalid options
- [x] Double check expected dynamic libraries check
- [ ] Add unit tests in addition to e2e?
- [ ] Release vs debug mode?
- [x] Make ic-wasm release and use the released crate

[LANG-124]: https://dfinity.atlassian.net/browse/LANG-124?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .github/workflows/publish.yml
Comment (kentosugama) [inline .github/workflows/publish.yml:84]: Is it okay to expect these libraries?

File: .github/workflows/publish.yml
Comment (kentosugama) [inline .github/workflows/publish.yml:96]: Ditto

File: src/dfx/src/lib/builders/mod.rs
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/builders/mod.rs:536]: please DRY
```suggestion
    let level = match level {
        // O3 and Oz empirically give best cycle savings and code size savings respectively
        WasmOptLevel::Cycles => WasmOptlevel::O3,
        WasmOptLevel::Size => WasmOptLevel::Oz,
        other => other,
    };
    ic_wasm::shrink::shrink_with_wasm_opt(&mut module, &level.to_string())
        .context("Could not optimize the WASM module.")?;
```

File: src/dfx/Cargo.toml
Comment (ericswanson-dfinity) [inline src/dfx/Cargo.toml:60]: Is this release imminent?  Or will we be holding this PR open for some time while we await ic-wasm 0.3.6?

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

File: src/dfx/Cargo.toml
Comment (kentosugama) [reply src/dfx/Cargo.toml:60]: Release is imminent. I was just asking Yan to hold off on it while this PR was being worked on in case some changes needed to be made

File: src/dfx-core/src/config/model/dfinity.rs
Comment (ericswanson-dfinity) [inline src/dfx-core/src/config/model/dfinity.rs:85]: Two concerns with using `serde alias`:
1. Any code that rewrites dfx.json using these types would  read `"cycles"` and write `"O3"`.  We don't have any such code now (ever since we removed `dfx config`), but if we add any in the future we will have to keep this in mind
2. The JSON schema for WasmOptLevel doesn't include the valid enum values of `cycles` and `size`.

My suggestion would be to re-add the `Cycles` and `Size` enum kinds, and transform them to `O3` and `Oz` in one place in the code rather than through serde.

File: .github/workflows/publish.yml
Comment (ericswanson-dfinity) [reply .github/workflows/publish.yml:84]: Yes, these are fine

**Review by ericswanson-dfinity — APPROVED** [review]
Looks good.  Approved assuming the update to ic-wasm 0.3.6 is applied.  The rust e2e test failures will be fixed once https://github.com/dfinity/sdk/pull/3099 is merged.

**Comment by kentosugama** [comment]
@ericswanson-dfinity Thank you for the reviews!

**Pull request merged**
